lexer/stateful/pattern.go

```diff
@@ -0,0 +1,218 @@
+// since otherwise it would take too much memory.
+const (
+	charclassSizeLimit = 4096
+	mayMatchZero       = 0
+	mayMatchOneOrMore  = 1
+	matchesExactlyOne  = 2
+)
+
+// Get all the possible starting points of a regexp that resolve to runes
+// Returns
+//   - the list of nodes that represent a potential starting point
+//   - an int indicating the match nature
+//       * a zero length match is possible
+//       * the match can be one or more characters
+//			 * the whole pattern matches only exactly one character
+func getPotentialStarts(r *syntax.Regexp) ([]*syntax.Regexp, int) {
+	var (
+		result []*syntax.Regexp
+	)
+
+	switch r.Op {
+	// Disjunction. The match nature returned is by priority ; zero-match, one-or-more, exactly one.
+	case syntax.OpAlternate:
+		var mayresult = matchesExactlyOne // will go downward with the arms that reply less
+		for _, reg := range r.Sub {
+			res, may := getPotentialStarts(reg)
+			mayresult = min(mayresult, may)
+			result = append(result, res...)
+		}
+		return result, mayresult
+
+	// A sequence. Checks that it has at least one non-zero length pattern and always returns one or more.
+	case syntax.OpConcat:
+		for _, reg := range r.Sub {
+			res, may := getPotentialStarts(reg)
+			result = append(result, res...)
+			if may != mayMatchZero {
+				return result, mayMatchOneOrMore
+			}
+		}
+		return result, mayMatchZero
+
+	case syntax.OpBeginLine,
+		syntax.OpBeginText,
+		syntax.OpWordBoundary,
+		syntax.OpNoWordBoundary,
+		syntax.OpNoMatch,
+		syntax.OpEmptyMatch,
+		syntax.OpEndLine,
+		syntax.OpEndText:
+		// We just ignore those as being "irrelevant" in the search for runes, as they are just "helpers"
+		// to perform certain cases of matches.
+		// Moreover, they're zero-length matches, which by itself will trigger an error for the regexp analysis
+		// if unused along stuff that actually consume runes.
+		return []*syntax.Regexp{}, mayMatchZero
+
+	case syntax.OpQuest, syntax.OpStar:
+		res, _ := getPotentialStarts(r.Sub[0])
+		return res, mayMatchZero
+
+	case syntax.OpRepeat:
+		res, nb := getPotentialStarts(r.Sub[0])
+		if r.Min == 0 {
+			return res, mayMatchZero
+		} else if r.Min == 1 && r.Max == 1 {
+			return res, nb
+		}
+		return res, mayMatchOneOrMore
+
+	case syntax.OpCapture:
+		return getPotentialStarts(r.Sub[0])
+
+	case syntax.OpPlus:
+		var nodes, nb = getPotentialStarts(r.Sub[0])
+		return nodes, min(nb, mayMatchOneOrMore) // (^\b+) is a potential zero length match
+
+	case syntax.OpLiteral:
+		// Literal can also visibly be a sequence of characters, hence the check on the length
+		// of the Runes.
+		res := []*syntax.Regexp{r}
+		if len(r.Rune) > 1 {
+			return res, mayMatchOneOrMore
+		}
+		return res, matchesExactlyOne
+
+	case syntax.OpCharClass, syntax.OpAnyChar, syntax.OpAnyCharNotNL:
+		// These are one character matches.
+		return []*syntax.Regexp{r}, matchesExactlyOne
+	}
+
+	panic("should not get here")
```

Comment:
I would typically also include contextual information in the panic, otherwise it can be very difficult to debug in "real life".